File: flutter/lib/src/leaderboard_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'leaderboard_config.dart';
import 'leaderboard_models.dart';
import 'leaderboard_name_policy.dart';
import 'leaderboard_storage.dart';

class LeaderboardClient {
  LeaderboardClient({
    required this.config,
    LeaderboardStorage? storage,
    LeaderboardNamePolicy? namePolicy,
    http.Client? httpClient,
  })  : storage = storage ?? const SharedPreferencesLeaderboardStorage(),
        namePolicy = namePolicy ??
            LeaderboardNamePolicy(maxLength: config.nameMaxLength),
        _httpClient = httpClient ?? http.Client();

  final LeaderboardConfig config;
  final LeaderboardStorage storage;
  final LeaderboardNamePolicy namePolicy;
  final http.Client _httpClient;

  LeaderboardSnapshot? _cachedSnapshot;
  DateTime? _cachedAt;
  String? _cachedPeriod;

  String get _playerIdKey =>
      '${config.namespace}.${config.scope}.leaderboard.playerId.v1';
  String get _playerNameKey =>
      '${config.namespace}.${config.scope}.leaderboard.playerName.v1';
  String get _legacyPlayerIdKey =>
      '${config.namespace}.leaderboard.playerId.v1';
  Future<String> get currentPeriod => _rpc('leaderboard_period', {});

  Future<String> _rpc(String action, Map<String, dynamic> params) async {
    if (config.usesEndpoint) {
      final payload = await _requestEndpointJson(method: 'POST', body: {
        'action': action, 'namespace': config.namespace, 'scope': config.scope, ...params,
      });
      final value = payload?[action == 'leaderboard_name' ? 'playerName' : 'period'];
      if (value is! String || value.isEmpty) {
        throw const LeaderboardException('Endpoint must implement name reservation and ranking periods.');
      }
      return value;
    }
    final response = await _httpClient.post(
      Uri.parse('${config.projectUrl}/rest/v1/rpc/$action'), headers: _headers,
      body: jsonEncode({'p_app_id': config.namespace, 'p_scope': config.scope, ...params}),
    ).timeout(config.requestTimeout);
    _requireSuccess(response);
    return jsonDecode(response.body) as String;
  }

  Map<String, String> get _headers => {
        'apikey': config.requiredAnonKey,
        'Authorization': 'Bearer ${config.requiredAnonKey}',
        'Content-Type': 'application/json',
      };

  Map<String, String> get _endpointHeaders => {
        'Content-Type': 'application/json',
      };

  Uri _tableUri([Map<String, String>? query]) {
    return Uri.parse(
      '${config.projectUrl}/rest/v1/${config.table}',
    ).replace(queryParameters: query);
  }

  Uri _endpointUri([Map<String, String>? query]) {
    if (config.endpoint == null || config.endpoint!.trim().isEmpty) {
      throw const LeaderboardConfigException('endpoint is not configured.');
    }
    return Uri.parse(config.endpoint!).replace(queryParameters: query);
  }

  Future<String> getOrCreatePlayerId() async {
    final saved = await storage.read(_playerIdKey);
    if (saved != null && saved.isNotEmpty) return saved;
    final legacy = config.scope == 'default'
        ? await storage.read(_legacyPlayerIdKey)
        : null;
    if (legacy != null && legacy.isNotEmpty) {
      await storage.write(_playerIdKey, legacy);
      return legacy;
    }
    final id = createAnonymousPlayerId();
    await storage.write(_playerIdKey, id);
    return id;
  }

  Future<String> getOrCreatePlayerName() async {
    final name = await _rpc('leaderboard_name', {
      (config.usesEndpoint ? 'playerId' : 'p_player_id'): await getOrCreatePlayerId(),
    });
    await storage.write(_playerNameKey, name);
    return name;
  }

  Future<String?> resolveCountryCode() async {
    return normalizeCountryCode(await config.countryCodeResolver?.call());
  }

  // Compatibility method: visiting a board must never submit an activation score.
  Future<void> activateCurrentPeriod() async { await currentPeriod; }

  Future<void> submitScore(int totalScore, {DateTime? completedAt}) async {
    final playedMonth = (completedAt ?? DateTime.now()).toUtc().toIso8601String().substring(0, 7);
    final period = await currentPeriod;
    if (playedMonth != period) throw const LeaderboardException('Gameplay belongs to an expired ranking period.');
    if (totalScore < 0) {
      throw const LeaderboardException('Score cannot be negative.');
    }
    if (config.usesEndpoint) {
      await _upsert(score: totalScore, period: period);
      await _rememberScore(period, totalScore);
      return;
    }

    final playerId = await getOrCreatePlayerId();
    final rows = await _requestList(
      _tableUri({
        'select': 'score',
        'app_id': 'eq.${config.namespace}',
        'scope': 'eq.${config.scope}',
        'player_id': 'eq.$playerId',
        'period': 'eq.$period',
        'limit': '1',
      }),
    );
    final previous = rows.isEmpty ? null : rows.first['score'] as num?;
    if (!config.isBetterScore(totalScore, previous)) return;
    await _upsert(score: totalScore, period: period);
    await _rememberScore(period, totalScore);
  }

  Future<String> updatePlayerName(String rawName) async {
    String name;
    try {
      name = namePolicy.sanitize(rawName);
    } on FormatException catch (error) {
      throw LeaderboardNameException(error.message);
    }

    final playerId = await getOrCreatePlayerId();

    if (config.usesEndpoint) {
      final payload = await _requestEndpointJson(
        method: 'PATCH',
        body: {
          'namespace': config.namespace,
          'playerId': playerId,
          'playerName': name,
          'scope': config.scope,
        },
      );
      final savedName = payload?['playerName'] as String? ?? name;
      await storage.write(_playerNameKey, savedName);
      invalidateCache();
      return savedName;
    }

    await _rpc('leaderboard_rename', {'p_player_id': playerId, 'p_name': name});
    await storage.write(_playerNameKey, name);
    invalidateCache();
    return name;
  }

  Future<LeaderboardSnapshot> fetchSnapshot({bool forceRefresh = false}) async {
    if (config.usesEndpoint) {
      return _fetchEndpointSnapshot(forceRefresh: forceRefresh);
    }

    final period = await currentPeriod;
    if (!forceRefresh &&
        _cachedSnapshot != null &&
        _cachedPeriod == period &&
        _cachedAt != null &&
        DateTime.now().difference(_cachedAt!) < config.cacheDuration) {
      return _cachedSnapshot!;
    }

    final rows = await _requestList(
      _tableUri({
        'select': 'player_id,name,score,country_code,updated_at',
        'app_id': 'eq.${config.namespace}',
        'scope': 'eq.${config.scope}',
        'period': 'eq.$period',
        'order':
            'score.${config.scoreOrder == LeaderboardScoreOrder.lower ? 'asc' : 'desc'},updated_at.desc',
        'limit': '${config.rankScanLimit}',
      }),
    );
    final ranked = [
      for (var index = 0; index < rows.length; index++)
        LeaderboardEntry.fromJson(rows[index], rank: index + 1),
    ];
    final playerId = await getOrCreatePlayerId();
    final current =
        ranked.where((entry) => entry.playerId == playerId).firstOrNull;
    final personal = await _personalRecord(period);
    final snapshot = LeaderboardSnapshot(
      entries: ranked.take(config.topLimit).toList(growable: false),
      currentPlayer: current,
      previousScore: personal['previous'] as int?,
      showPrevious: current != null && personal['previous'] != null && config.isBetterScore(personal['previous'] as int, current.score),
    );
    if (snapshot.currentPlayer != null) await _rememberScore(period, snapshot.currentPlayer!.score);
    _cachedSnapshot = snapshot;
    _cachedAt = DateTime.now();
    _cachedPeriod = period;
    return snapshot;
  }

  Future<Map<String, dynamic>> _personalRecord(String period) async {
    final key = '${config.namespace}.${config.scope}.${await getOrCreatePlayerId()}.personal.v1';
    final raw = await storage.read(key);
    final state = raw == null ? <String, dynamic>{'period': period, 'current': null, 'previous': null} : jsonDecode(raw) as Map<String, dynamic>;
    if (state['period'] != period) {
      if (state['current'] != null && config.isBetterScore(state['current'] as int, state['previous'] as num?)) state['previous'] = state['current'];
      state['current'] = null;
      state['period'] = period;
    }
    state['key'] = key;
    await storage.write(key, jsonEncode(state));
    return state;
  }

  Future<void> _rememberScore(String period, int score) async {
    final state = await _personalRecord(period);
    if (config.isBetterScore(score, state['current'] as num?)) state['current'] = score;
    await storage.write(state['key'] as String, jsonEncode(state));
  }

  void invalidateCache() {
    _cachedSnapshot = null;
    _cachedAt = null;
    _cachedPeriod = null;
  }

  Future<void> _upsert({required int score, required String period}) async {
    if (config.usesEndpoint) {
      await _requestEndpointJson(
        method: 'POST',
        body: {
          'countryCode': await resolveCountryCode(),
          'namespace': config.namespace,
          'playerId': await getOrCreatePlayerId(),
          'playerName': await getOrCreatePlayerName(),
          'period': period,
          'score': score,
          'scope': config.scope,
        },
      );
      invalidateCache();
      return;
    }

    final response = await _httpClient
        .post(
          _tableUri({'on_conflict': 'app_id,scope,player_id,period'}),
          headers: {
            ..._headers,
            'Prefer': 'resolution=merge-duplicates,return=minimal',
          },
          body: jsonEncode({
            'app_id': config.namespace,
            'scope': config.scope,
            'player_id': await getOrCreatePlayerId(),
            'name': await getOrCreatePlayerName(),
            'country_code': await resolveCountryCode(),
            'score': score,
            'period': period,
            'updated_at': DateTime.now().toUtc().toIso8601String(),
          }),
        )
        .timeout(config.requestTimeout);
    _requireSuccess(response);
    invalidateCache();
  }

  Future<List<Map<String, dynamic>>> _requestList(Uri uri) async {
    final response = await _httpClient
        .get(uri, headers: _headers)
        .timeout(config.requestTimeout);
    _requireSuccess(response);
    final decoded = jsonDecode(response.body);
    if (decoded is! List) {
      throw const LeaderboardException('Leaderboard returned invalid data.');
    }
    return decoded.cast<Map<String, dynamic>>();
  }

  Future<Map<String, dynamic>?> _requestEndpointJson({
    required String method,
    Map<String, String>? query,
    Map<String, dynamic>? body,
  }) async {
    final uri = _endpointUri(query);
    final encodedBody = body == null ? null : jsonEncode(body);
    final response = switch (method) {
      'GET' => await _httpClient
          .get(uri, headers: _endpointHeaders)
          .timeout(config.requestTimeout),
      'PATCH' => await _httpClient
          .patch(uri, headers: _endpointHeaders, body: encodedBody)
          .timeout(config.requestTimeout),
      'POST' => await _httpClient
          .post(uri, headers: _endpointHeaders, body: encodedBody)
          .timeout(config.requestTimeout),
      _ => throw LeaderboardException('Unsupported endpoint method: $method.'),
    };
    _requireSuccess(response);
    if (response.body.trim().isEmpty) return null;
    final decoded = jsonDecode(response.body);
    if (decoded is! Map<String, dynamic>) {
      throw const LeaderboardException('Leaderboard returned invalid data.');
    }
    return decoded;
  }

  Future<LeaderboardSnapshot> _fetchEndpointSnapshot({
    bool forceRefresh = false,
  }) async {
    final period = await currentPeriod;
    if (!forceRefresh &&
        _cachedSnapshot != null &&
        _cachedPeriod == period &&
        _cachedAt != null &&
        DateTime.now().difference(_cachedAt!) < config.cacheDuration) {
      return _cachedSnapshot!;
    }

    final payload = await _requestEndpointJson(
      method: 'GET',
      query: {
        'limit': '${config.topLimit}',
        'namespace': config.namespace,
        'playerId': await getOrCreatePlayerId(),
        'scope': config.scope,
      },
    );
    final entriesJson = payload?['entries'];
    if (entriesJson is! List) {
      throw const LeaderboardException('Leaderboard returned invalid data.');
    }
    final entries = [
      for (final entry in entriesJson)
        if (entry is Map<String, dynamic>)
          LeaderboardEntry.fromEndpointJson(entry),
    ];
    final currentJson = payload?['currentPlayer'] ?? payload?['playerEntry'];
    final currentPlayer = currentJson is Map<String, dynamic>
        ? LeaderboardEntry.fromEndpointJson(currentJson)
        : null;
    final personal = await _personalRecord(period);
    final snapshot = LeaderboardSnapshot(
      entries: entries,
      currentPlayer: currentPlayer,
      previousScore: personal['previous'] as int?,
      showPrevious: currentPlayer != null && personal['previous'] != null && config.isBetterScore(personal['previous'] as int, currentPlayer.score),
    );
    if (snapshot.currentPlayer != null) await _rememberScore(period, snapshot.currentPlayer!.score);
    _cachedSnapshot = snapshot;
    _cachedAt = DateTime.now();
    _cachedPeriod = period;
    return snapshot;
  }

  void _requireSuccess(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw LeaderboardException(
        'Leaderboard request failed (${response.statusCode}).',
        statusCode: response.statusCode,
      );
    }
  }

  String _escapeFilter(String value) {
    return value
        .replaceAll('\\', '\\\\')
        .replaceAll('%', '\\%')
        .replaceAll('_', '\\_');
  }
}

extension<T> on Iterable<T> {
  T? get firstOrNull {
    final iterator = this.iterator;
    return iterator.moveNext() ? iterator.current : null;
  }
}

File: react/src/leaderboardClient.js
import { createNamePolicy } from "./namePolicy.js";
import {
  createAnonymousPlayerId,
  isBetterScore,
  normalizeCountryCode,
  normalizeScoreOrder,
} from "./utils.js";

export function createLeaderboardClient({
  endpoint,
  supabaseUrl,
  supabaseAnonKey,
  table = "app_leaderboard_scores",
  namespace,
  scope = "default",
  scoreOrder = "higher",
  activationScore = 0,
  topLimit = 100,
  rankScanLimit = 500,
  cacheDurationMs = 45_000,
  requestTimeoutMs = 8_000,
  countryCodeResolver,
  storage = globalThis.localStorage,
  namePolicy = createNamePolicy(),
  fetchImpl = globalThis.fetch,
}) {
  if (!endpoint && (!supabaseUrl || !supabaseAnonKey)) {
    throw new Error(
      "Either endpoint or supabaseUrl and supabaseAnonKey are required.",
    );
  }
  if (!namespace) throw new Error("namespace is required.");
  if (!scope) throw new Error("scope is required.");
  if (!storage) throw new Error("A storage implementation is required.");
  if (!fetchImpl) throw new Error("A fetch implementation is required.");

  const projectUrl = supabaseUrl
    ?.trim()
    .replace(/\/rest\/v1\/?$/, "")
    .replace(/\/$/, "");
  const endpointUrl = endpoint?.trim().replace(/\/$/, "");
  const normalizedScoreOrder = normalizeScoreOrder(scoreOrder);
  const storagePrefix = `${namespace}.${scope}.leaderboard`;
  const playerIdKey = `${storagePrefix}.playerId.v1`;
  const playerNameKey = `${storagePrefix}.playerName.v1`;
  const legacyPlayerIdKey = `${namespace}.leaderboard.playerId.v1`;
  let cache = null;

  const headers = {
    apikey: supabaseAnonKey,
    Authorization: `Bearer ${supabaseAnonKey}`,
    "Content-Type": "application/json",
  };

  async function rpc(action, params) {
    if (endpointUrl) {
      const payload = await requestEndpointJson({}, {
        method: "POST", body: JSON.stringify({ action, namespace, scope, ...params }),
      });
      const value = payload?.[action === "leaderboard_name" ? "playerName" : "period"];
      if (typeof value !== "string" || !value) throw new Error("Endpoint must implement name reservation and ranking periods.");
      return value;
    }
    const response = await request(`${projectUrl}/rest/v1/rpc/${action}`, {
      method: "POST", body: JSON.stringify({ p_app_id: namespace, p_scope: scope, ...params }),
    });
    return response.json();
  }

  async function currentPeriod() {
    return rpc("leaderboard_period", {});
  }

  function tableUrl(params = {}) {
    const url = new URL(`${projectUrl}/rest/v1/${table}`);
    Object.entries(params).forEach(([key, value]) => {
      url.searchParams.set(key, value);
    });
    return url;
  }

  function endpointUrlWithParams(params = {}) {
    const url = new URL(
      endpointUrl,
      globalThis.location?.origin ?? "http://localhost",
    );
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        url.searchParams.set(key, value);
      }
    });
    return url;
  }

  async function request(url, options = {}) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), requestTimeoutMs);
    try {
      const response = await fetchImpl(url, {
        ...options,
        headers: {
          ...(endpointUrl ? { "Content-Type": "application/json" } : headers),
          ...options.headers,
        },
        signal: controller.signal,
      });
      if (!response.ok) {
        throw new Error(`Leaderboard request failed (${response.status}).`);
      }
      return response;
    } finally {
      clearTimeout(timer);
    }
  }

  async function requestList(params) {
    const response = await request(tableUrl(params));
    const rows = await response.json();
    if (!Array.isArray(rows)) {
      throw new Error("Leaderboard returned invalid data.");
    }
    return rows;
  }

  async function requestEndpointJson(params = {}, options = {}) {
    const response = await request(endpointUrlWithParams(params), options);
    const text = await response.text();
    if (!text.trim()) return null;
    return JSON.parse(text);
  }

  function getOrCreatePlayerId() {
    const saved = storage.getItem(playerIdKey);
    if (saved) return saved;
    const legacy =
      scope === "default" ? storage.getItem(legacyPlayerIdKey) : null;
    if (legacy) {
      storage.setItem(playerIdKey, legacy);
      return legacy;
    }
    const id = createAnonymousPlayerId();
    storage.setItem(playerIdKey, id);
    return id;
  }

  async function getOrCreatePlayerName() {
    const name = await rpc("leaderboard_name", endpointUrl
      ? { playerId: getOrCreatePlayerId() }
      : { p_player_id: getOrCreatePlayerId() });
    storage.setItem(playerNameKey, name);
    return name;
  }

  async function resolveCountryCode() {
    return normalizeCountryCode(await countryCodeResolver?.());
  }

  async function upsert(score, period) {
    if (endpointUrl) {
      await requestEndpointJson(
        {},
        {
          method: "POST",
          body: JSON.stringify({
            period,
            countryCode: await resolveCountryCode(),
            namespace,
            playerId: getOrCreatePlayerId(),
            playerName: await getOrCreatePlayerName(),
            score,
            scope,
          }),
        },
      );
      invalidateCache();
      return;
    }

    await request(
      tableUrl({ on_conflict: "app_id,scope,player_id,period" }),
      {
        method: "POST",
        headers: { Prefer: "resolution=merge-duplicates,return=minimal" },
        body: JSON.stringify({
          app_id: namespace,
          player_id: getOrCreatePlayerId(),
          period,
          scope,
          name: await getOrCreatePlayerName(),
          country_code: await resolveCountryCode(),
          score,
          updated_at: new Date().toISOString(),
        }),
      },
    );
    invalidateCache();
  }

  // Kept for source compatibility. Opening a board never creates a score.
  async function activateCurrentPeriod() { await currentPeriod(); }

  async function submitScore(totalScore, { completedAt = new Date().toISOString() } = {}) {
    const playedMonth = new Date(completedAt).toISOString().slice(0, 7);
    const period = await currentPeriod();
    if (playedMonth !== period) throw new Error("Gameplay belongs to an expired ranking period.");
    if (!Number.isSafeInteger(totalScore) || totalScore < 0) {
      throw new Error("Score must be a non-negative integer.");
    }
    if (endpointUrl) {
      await upsert(totalScore, period);
      rememberScore(period, totalScore);
      return;
    }

    const rows = await requestList({
      select: "score",
      app_id: `eq.${namespace}`,
      scope: `eq.${scope}`,
      player_id: `eq.${getOrCreatePlayerId()}`,
      period: `eq.${period}`,
      limit: "1",
    });
    if (!isBetterScore(totalScore, rows[0]?.score, normalizedScoreOrder)) return;
    await upsert(totalScore, period);
      rememberScore(period, totalScore);
  }

  async function updatePlayerName(rawName) {
    const name = namePolicy.sanitize(rawName);
    const playerId = getOrCreatePlayerId();

    if (endpointUrl) {
      const payload = await requestEndpointJson(
        {},
        {
          method: "PATCH",
          body: JSON.stringify({
            namespace,
            playerId,
            playerName: name,
            scope,
          }),
        },
      );
      const savedName = payload?.playerName ?? name;
      storage.setItem(playerNameKey, savedName);
      invalidateCache();
      return savedName;
    }

    await rpc('leaderboard_rename', { p_player_id: playerId, p_name: name });
    storage.setItem(playerNameKey, name);
    invalidateCache();
    return name;
  }

  async function fetchSnapshot({ forceRefresh = false } = {}) {
    if (endpointUrl) {
      return fetchEndpointSnapshot({ forceRefresh });
    }

    const period = await currentPeriod();
    if (
      !forceRefresh &&
      cache?.period === period &&
      Date.now() - cache.createdAt < cacheDurationMs
    ) {
      return cache.snapshot;
    }

    const rows = await requestList({
      select: "player_id,name,score,country_code,updated_at",
      app_id: `eq.${namespace}`,
      scope: `eq.${scope}`,
      period: `eq.${period}`,
      order: `score.${normalizedScoreOrder === "lower" ? "asc" : "desc"},updated_at.desc`,
      limit: String(rankScanLimit),
    });
    const ranked = rows.map((row, index) => ({
      rank: index + 1,
      playerId: row.player_id ?? "",
      name: row.name?.trim() || "Player",
      score: Number(row.score) || 0,
      countryCode: normalizeCountryCode(row.country_code),
      updatedAt: row.updated_at,
    }));
    const snapshot = {
      entries: ranked.slice(0, topLimit),
      currentPlayer:
        ranked.find((row) => row.playerId === getOrCreatePlayerId()) ?? null,
    };
    decorateSnapshot(snapshot, period);
    cache = { period, createdAt: Date.now(), snapshot };
    return snapshot;
  }

  async function fetchEndpointSnapshot({ forceRefresh = false } = {}) {
    const period = await currentPeriod();
    if (
      !forceRefresh &&
      cache?.period === period &&
      Date.now() - cache.createdAt < cacheDurationMs
    ) {
      return cache.snapshot;
    }

    const payload = await requestEndpointJson({
      limit: String(topLimit),
      namespace,
      playerId: getOrCreatePlayerId(),
      scope,
    });
    if (!payload || !Array.isArray(payload.entries)) {
      throw new Error("Leaderboard returned invalid data.");
    }
    const mapEntry = (entry) => ({
      rank: Number(entry.rank) || 0,
      playerId: entry.playerId ?? entry.player_id ?? "",
      name: entry.name ?? entry.playerName ?? entry.player_name ?? "Player",
      score: Number(entry.score ?? entry.elapsedSeconds) || 0,
      countryCode: normalizeCountryCode(
        entry.countryCode ?? entry.country_code,
      ),
      updatedAt: entry.updatedAt ?? entry.updated_at ?? entry.submittedAt,
    });
    const snapshot = {
      entries: payload.entries.map(mapEntry),
      currentPlayer: payload.currentPlayer
        ? mapEntry(payload.currentPlayer)
        : payload.playerEntry
          ? mapEntry(payload.playerEntry)
          : null,
    };
    decorateSnapshot(snapshot, period);
    cache = { period, createdAt: Date.now(), snapshot };
    return snapshot;
  }

  function personalRecord(period) {
    const key = `${storagePrefix}.${getOrCreatePlayerId()}.personal.v1`;
    const state = JSON.parse(storage.getItem(key) || 'null') || { period, current: null, previous: null };
    if (state.period !== period) {
      if (state.current !== null && isBetterScore(state.current, state.previous, normalizedScoreOrder)) state.previous = state.current;
      state.current = null;
      state.period = period;
    }
    return { key, state };
  }
  function rememberScore(period, score) {
    const { key, state } = personalRecord(period);
    if (isBetterScore(score, state.current, normalizedScoreOrder)) state.current = score;
    storage.setItem(key, JSON.stringify(state));
  }
  function decorateSnapshot(snapshot, period) {
    const { key, state } = personalRecord(period);
    if (snapshot.currentPlayer && isBetterScore(snapshot.currentPlayer.score, state.current, normalizedScoreOrder)) state.current = snapshot.currentPlayer.score;
    storage.setItem(key, JSON.stringify(state));
    snapshot.previousScore = state.previous;
    snapshot.showPrevious = !!snapshot.currentPlayer && state.previous !== null &&
      isBetterScore(state.previous, snapshot.currentPlayer.score, normalizedScoreOrder);
    return snapshot;
  }

  function invalidateCache() {
    cache = null;
  }

  return {
    activateCurrentPeriod,
    currentPeriod,
    fetchSnapshot,
    fetchEndpointSnapshot,
    getOrCreatePlayerId,
    getOrCreatePlayerName,
    invalidateCache,
    submitScore,
    updatePlayerName,
  };
}
